### DISCUSSION TITLE
Unpatch / two-way JSON diff

### DISCUSSION THREAD

**Discussion by khiner** [nlohmann/json#3396] [Q&A]
Is there an easy way to generate the inverse diff from a given `json` diff, if I only have access to the diff itself, and not the original `json` objects that created the diff?

Let's say I have:

```cpp
json old_json = ...;
json new_json = ...;
json diff = json::diff(old_json, new_json);
```

Here I could make the inverse diff with `json::diff(new_json, old_json)`, but suppose I don't have access to these at some later point, and only have the `diff`, and want to transform it into its own inverse, which would be equivalent to the result of `json::diff(new_json, old_json)`. It seems like all the information needed to do so is there, but might be a bit tricky to get all the cases right.

Is anyone aware of an existing solution for this?

Thanks!

**Comment by nlohmann** [comment]
I am not aware of a solution, but maybe I have not fully understood what you want.

Can you provide a small example?

**Reply by khiner** [reply]
Thanks for getting back to me. I can definitely elaborate. Typing it out actually helped me realize the thing I wanted isn't actually well posed, since Diffs or a Patch won't have the information needed for any function to generate its inverse. I'll explain more below and then get to a more clear explanation of the functionality I could use help implementing well.

Take the [JSON Pointer and JSON Patch](https://github.com/nlohmann/json#json-pointer-and-json-patch) in the README as an example:

```cpp
// a JSON value
json j_original = R"({
  "baz": ["one", "two", "three"],
  "foo": "bar"
})"_json;

// access members with a JSON pointer (RFC 6901)
j_original["/baz/1"_json_pointer];
// "two"

// a JSON patch (RFC 6902)
json j_patch = R"([
  { "op": "replace", "path": "/baz", "value": "boo" },
  { "op": "add", "path": "/hello", "value": ["world"] },
  { "op": "remove", "path": "/foo"}
])"_json;

// apply the patch
json j_result = j_original.patch(j_patch);
// {
//    "baz": "boo",
//    "hello": ["world"]
// }

// calculate a JSON patch from two JSON values
json::diff(j_result, j_original);
// [
//   { "op":" replace", "path": "/baz", "value": ["one", "two", "three"] },
//   { "op": "remove","path": "/hello" },
//   { "op": "add", "path": "/foo", "value": "bar" }
// ]
```

Given the final output of this example,

```cpp
const Diffs diffs{[
  { "op":" replace", "path": "/baz", "value": ["one", "two", "three"] },
  { "op": "remove","path": "/hello" },
  { "op": "add", "path": "/foo", "value": "bar" }
]};
```

I am looking for a function like an `inverse_diff(const Diffs &diff)`, which produces the same thing as ` json::diff(j_original, j_result);` in this example. So, if given diffs as input, it would return the original patch in this case:

```cpp
json j_patch = R"([
  { "op": "replace", "path": "/baz", "value": "boo" },
  { "op": "add", "path": "/hello", "value": ["world"] },
  { "op": "remove", "path": "/foo"}
])"_json;
```

As I mentioned above, there obviously couldn't be such a function. I understand better though now what I am looking for, explained below.

## Use case
I am developing a generic undo/redo system for an application's state. The state is a struct that is json-convertible. In order to not have to write both a forward and reverse (do/redo & undo) version of each state-modifying action, the actions in my application simply modify the state directly.

On application startup, after the default state is initialized, it is copied to a json variable stored in the main context. After each action that modifies the main state struct, I basically do:

```cpp
const Diffs &forward_diff = json::diff(old_state_json, new_state_json);
const Diffs &reverse_diff = json::diff(new_state_json, old_state_json);
```

and then store the forward/reverse pair on the undo stack:

```cpp
undo_stack.push(ActionDiffs{ forward_diff, reverse_diff });
```

## Need

The problem with this approach is that the state can be a very large object, and it's important that this process is as fast as possible for the application to be snappy.

For something like e.g. changing a single bool value in the whole struct, it's pretty inefficient to calculate

```cpp
const Diffs &forward_diff = json::diff(old_state_json, new_state_json);
const Diffs &reverse_diff = json::diff(new_state_json, old_state_json);
```

For something like this, two full passes is pretty silly, since we could generate some diff-like object with both the change from e.g. true to false, and the false to true inverse change in one pass. So this is what I'm looking for - some function like

```cpp
std::pair<Diffs, Diffs> {forward_diff, reverse_diff} = json::diff_with_inverse(old_state_json, new_state_json);
```

I'm not sure how common this kind of poor-man's persistent data structure need is, but seems like it would be handy for others with similar needs in e.g. the reactive programming context.

Is this, or something like it, you could ever see making it into the library? (Happy to help, but it would be my first time contributing.)

Thank you for your time and attention!

**Reply by falbrechtskirchinger** [reply]
I would call that a two-way diff. Sounds interesting and like something I could see myself use. Don't know how feasible an efficient algorithm to compute such a diff is but I wish you best of luck should you choose to pursue a PR. The current diff algorithm isn't exactly great either and could definitely be improved.

**Reply by nlohmann** [reply]
The current diff algorithm is indeed a placeholder for any better approach ;-)

**Reply by khiner** [reply]
Thanks all. When my application state actually does become large enough for this to be a noticeable performance issue, I will consider both improving the existing diff algorithm, and taking a stab at a two-way diff algorithm. In JS-land, I've used the `unpatch` feature of [jsondiffpatch](https://github.com/benjamine/jsondiffpatch#usage) for a case very similar to this.

Note that jsondiffpatch also supports using [Google's diff-match-patch](https://github.com/google/diff-match-patch) for more space-efficient patching of long strings. I currently happen to break out a couple long strings in my state into separate fields and use [this stl implementation of diff-match-patch](https://github.com/leutloff/diff-match-patch-cpp-stl) in a similar two-way approach as the json method above. It would be very convenient for me to just keep long text strings in my json and support a flag for `diff-match-patch`-diffing/patching on long strings. However, that would require a new `Diffs` structure (perhaps similar to [jsondiffpatch's delta format](https://github.com/benjamine/jsondiffpatch/blob/master/docs/deltas.md)).

This is really two separate issues (two-way diff algorithm/format & specialized patch format for long strings), so maybe this should be a separate topic. But just to get a feel on this in case I do get to implementing - how important is it to the maintainers that all patching is consistent with [rfc6902](https://datatracker.ietf.org/doc/html/rfc6902)? Would a custom diff-match-patch patch format addition for long strings (likely using a separate `patch` method) ever be considered in this package?

**Reply by khiner** [reply]
(Updated title for better searchability, using falbrechtskirchinger's nice "two-way diff" suggestion.)
